Implement integration tests for the federation messages receive feature

cmrd-senya opened the pull request:
This is some initial tests, more to come.

refs #5114 #6479

Commit:
Implement integration tests for the federation messages receive feature

These are some initial tests, more to come.

It tests some features of Request, StatusMessage, Comment, Like,
Participation, Retraction, SignedRetraction, RelayableRetraction entities
receive process.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+require 'spec_helper'
```
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+
+describe Workers::ReceiveEncryptedSalmon do
+  def generate_xml(entity)
+    xml = DiasporaFederation::Salmon::EncryptedSlap.generate_xml(@remote_user.diaspora_handle,
```
Useless assignment to variable - `xml`.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+describe Workers::ReceiveEncryptedSalmon do
+  def generate_xml(entity)
+    xml = DiasporaFederation::Salmon::EncryptedSlap.generate_xml(@remote_user.diaspora_handle,
+      OpenSSL::PKey::RSA.new(@remote_user.encryption_key),
```
Align the parameters of a method call if they span more than one line.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+  def generate_xml(entity)
+    xml = DiasporaFederation::Salmon::EncryptedSlap.generate_xml(@remote_user.diaspora_handle,
+      OpenSSL::PKey::RSA.new(@remote_user.encryption_key),
+      entity,
```
Align the parameters of a method call if they span more than one line.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+    xml = DiasporaFederation::Salmon::EncryptedSlap.generate_xml(@remote_user.diaspora_handle,
+      OpenSSL::PKey::RSA.new(@remote_user.encryption_key),
+      entity,
+      OpenSSL::PKey::RSA.new(@user.encryption_key)
```
Align the parameters of a method call if they span more than one line.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+      OpenSSL::PKey::RSA.new(@remote_user.encryption_key),
+      entity,
+      OpenSSL::PKey::RSA.new(@user.encryption_key)
+    )
```
Align `)` with `(`.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+
+  before do
+    @user = alice
+    allow(User).to receive(:find){ |id|
```
Space missing to the left of {.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+    allow(User).to receive(:find){ |id|
+      if id == @user.id
+        @user
+      else
```
Redundant `else`-clause.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+
+  it "doesn't save the status message if there is no sharing" do
+    entity = FactoryGirl.build(:status_message_entity,
+      diaspora_id: @remote_user.diaspora_handle,
```
Align the parameters of a method call if they span more than one line.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+  it "doesn't save the status message if there is no sharing" do
+    entity = FactoryGirl.build(:status_message_entity,
+      diaspora_id: @remote_user.diaspora_handle,
+      public: false
```
Align the elements of a hash literal if they span more than one line.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+    entity = FactoryGirl.build(:status_message_entity,
+      diaspora_id: @remote_user.diaspora_handle,
+      public: false
+    )
```
Align `)` with `(`.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+
+    it "treats status message receive correctly" do
+      entity = FactoryGirl.build(:status_message_entity,
+        diaspora_id: @remote_user.diaspora_handle,
```
Align the parameters of a method call if they span more than one line.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+    it "treats status message receive correctly" do
+      entity = FactoryGirl.build(:status_message_entity,
+        diaspora_id: @remote_user.diaspora_handle,
+        public: false
```
Align the elements of a hash literal if they span more than one line.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+      entity = FactoryGirl.build(:status_message_entity,
+        diaspora_id: @remote_user.diaspora_handle,
+        public: false
+      )
```
Align `)` with `(`.

diaspora-code-review commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,93 @@
+        diaspora_id: @remote_user.diaspora_handle,
+        public: false
+      )
+      
```
Trailing whitespace detected.

cmrd-senya commented:
This new version of the patch utilizes the latest changes in the diaspora_federation gem and does an integration test on comments receive. More tests to come.

cmrd-senya commented:
@jhass, this contains fix to the security issue with signatures we discovered before. I put it here, because here I test it using the integration tests.

I propose to merge it all together, what do you think?

There is to major things to do, before we can merge it:
- [ ] Do something to factories.rb in this patch
- [ ] Release the diaspora_federation gem with the latest changes

factories.rb here merged from the diaspora_federation gem and has some changes.

cmrd-senya commented:
@jhass, this contains fix to the security issue with signatures we discovered before. I put it here, because here I test it using the integration tests.

I propose to merge it all together, what do you think?

There is to major things to do, before we can merge it:
- [ ] Do something to factories.rb in this patch
- [ ] Release the diaspora_federation gem with the latest changes

factories.rb here merged from the diaspora_federation gem and has some changes. I think it is bad idea to duplicate code, so it would be nice if we could managed to reuse it somehow. Maybe we can ship this factory definitions with the diaspora_federation gem? What do you think, @SuperTux88?

jhass commented on config/initializers/diaspora_federation.rb:
```diff
@@ -61,5 +35,26 @@
     end
+
+    on :fetch_private_key_by_id do |diaspora_id|
+      Person.find_by(diaspora_handle: diaspora_id).owner.encryption_key
```
We shouldn't need to allocate AR objects here, the right combination of `pluck` and `joins` should do it.

jhass commented on config/initializers/diaspora_federation.rb:
```diff
@@ -61,5 +35,26 @@
+
+    on :fetch_private_key_by_post_guid do |guid|
+      author_user = Post.find_by(guid: guid).author.owner
+      author_user.encryption_key unless author_user.nil?
```
Same.

jhass commented on config/initializers/diaspora_federation.rb:
```diff
@@ -61,5 +35,26 @@
+    end
+
+    on :post_author_is_local? do |guid|
+      Post.find_by(guid: guid).author.local?
```
Is the post guid really all the information we have at this point in the gem? I think this hook is highly confusing, I'd prefer much much more to yield the diaspora id of the author. Actually same applies to the other two hooks.

cmrd-senya replied:
Well, yes, we know only parent post guid. Look at the [comment structure](https://github.com/SuperTux88/diaspora_federation/blob/salmon/lib/diaspora_federation/entities/relayable.rb). Instead I can propose to add a hook ```fetch_diaspora_id_by_post_guid``` then and call two hooks consequently to receive information about parent author.

jhass replied:
Mmh, doing two queries is worse, so I guess we keep it for now.

jhass commented on spec/integration/federation/factories.rb:
```diff
@@ -0,0 +1,219 @@
+
+def r_str
+  SecureRandom.hex(3)
+end
```
I don't think extracting this simple call into a method is doing any good.

jhass commented on spec/integration/federation/factories.rb:
```diff
@@ -0,0 +1,219 @@
+  klass.class_props.map { |prop|
+    [prop[:name], data[prop[:name]]] unless data[prop[:name]].nil?
+  }.compact.to_h
+end
```
The name of the method feels a bit too generic for this project. Also note that hashes compare order independent, `{a: 1, b: 2} ==  {b: 2, a: 1 } #=> true`

jhass commented on spec/integration/federation/factories.rb:
```diff
@@ -0,0 +1,219 @@
+  DiasporaFederation::Entities::Relayable.update_signatures!(
+    sort_hash(FactoryGirl.attributes_for(entity_type), FactoryGirl.factory_by_name(entity_type).build_class)
+  )
+end
```
The method seems unused?

jhass commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,224 @@
+require "spec_helper"
+require File.dirname(__FILE__) + "/factories"
```
Rather than this, place the factories into the correct folder, `spec/support/federation_integration_factories.rb` or something.

cmrd-senya replied:
If it is possible, I'd prefer to use factories shipped from the diaspora_federation gem, instead of copying it here.

jhass commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,224 @@
+                                              .with(
+                                                :save_person_after_webfinger,
+                                                kind_of(DiasporaFederation::Entities::Person)
+                                              ).and_call_original
```
I'm not sure why you do this. `allow` doesn't raise if the call doesn't happen and you don't return a mock.

cmrd-senya replied:
When DiasporaFederation.callbacks is called with some arguments which weren't specified by expect/allow it fails, that's why. I put expect for a different callback, and because of this all other original callbacks fail to call.

jhass replied:
So this should still be an `expect`?

cmrd-senya replied:
I just checked, it works without it now, because I use the same remote user everywhere. Earlier it was different.

jhass commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,224 @@
+                                              .and_call_original
+
+    Workers::ReceiveEncryptedSalmon.new.perform(user.id, generate_xml(entity, remote_user, user))
+    expect(klass.find_by(guid: entity.guid)).to be_nil
```
`expect(klass.exists?(...).to be_false` might be a bit clearer.

jhass commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,224 @@
+    Workers::ReceiveEncryptedSalmon.new.perform(user.id, generate_xml(entity, remote_user, user))
+    expect(klass.find_by(guid: entity.guid)).to be_nil
+  end
+end
```
This should probably go into its own file.

jhass commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,224 @@
+describe Workers::ReceiveEncryptedSalmon do
+  before do
+    @user = alice
+    allow(User).to receive(:find) { |id|
```
This again doesn't really do anything.

jhass commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,224 @@
+    entity = FactoryGirl.build(:request_entity, recipient_id: @user.diaspora_handle)
+
+    allow(Diaspora::Fetcher::Public).to receive(:queue_for)
+    expect(Diaspora::Fetcher::Public).to receive(:queue_for).exactly(1).times
```
No need for the `allow` here.

jhass commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,224 @@
+    Workers::ReceiveEncryptedSalmon.new.perform(@user.id, generate_xml(entity, @remote_user, @user))
+
+    expect(@user.contacts.count).to eq(2)
+    new_contact = @user.contacts.last
```
Not guaranteed to be stable, add an `order` clause.

jhass commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,224 @@
+      Workers::ReceiveEncryptedSalmon.new.perform(@user.id, generate_xml(entity, @remote_user, @user))
+
+      new_status_message = StatusMessage.find_by(guid: entity.guid)
+      expect(new_status_message).not_to be_nil
```
No real need for the local and again just use `exists?`

jhass commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,224 @@
+    Workers::ReceiveEncryptedSalmon.new.perform(@user.id, generate_xml(entity, @remote_user, @user))
+
+    new_status_message = StatusMessage.find_by(guid: entity.guid)
+    expect(new_status_message).to be_nil
```
Ditto.

jhass commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,224 @@
+        Workers::ReceiveEncryptedSalmon.new.perform(@user.id, generate_xml(remote_message, @remote_user, @user))
+      end
+
+      it_behaves_like "it deals correctly with a relayable" do
```
Let's wrap these into `context "with comment"` etc.

cmrd-senya commented on spec/integration/federation/receive_federation_messages.rb:
```diff
@@ -0,0 +1,216 @@
+    Workers::ReceiveEncryptedSalmon.new.perform(user.id, generate_xml(entity, remote_user, user))
+    received_entity = klass.find_by(guid: entity.guid)
+    expect(received_entity).not_to be_nil
+    expect(received_entity.author.diaspora_handle).to eq(remote_user.person.diaspora_handle)
```
@jhass, do I need to avoid AR object here?

jhass replied:
I'd say it's okay if you access multiple properties. It's more critical in non test code anyway.

cmrd-senya commented:
I'm kinda finished, it's left to make a release of the diaspora_federation.

It covers only a few cases of federation receive, but I think it's better to merge it and then add more tests in further commits, so we don't stay with the opened vulnerability.

jhass commented on config/initializers/diaspora_federation.rb:
```diff
@@ -61,5 +35,30 @@
+
+    on :post_author_is_local? do |guid|
+      id_array = Post.where(guid: guid).joins(:author).pluck(:owner_id)
+      !id_array.first.nil? unless id_array.empty?
```
Wouldn't something like `Post.joins(:author).where(guid: guid).not(owner_id: nil).exists?` work?

cmrd-senya replied:
Look at the [spec](https://github.com/cmrd-senya/diaspora/commit/cc821e19744ef57a8155763bb2e5bee06f0b2699#diff-89118032a5e2b114d8286b81e0db3b02R235). There is three different cases, and I made it to return nil if guid is not found. With what you offered it would be false in that case, which isn't true strictly.

jhass replied:
Then this shouldn't be a predicate (`foo?`), they should only return `true` or `false`, but rather `post_author_type` with the possible outcomes `:post_not_found`, `:local`, `:remote` and theoretically `:author_not_found`. The third possibility would be to raise in that case.

I think I do prefer to consider the post not found case `false`.

cmrd-senya replied:
Ok, that would be simpler then.

jhass commented on spec/integration/federation/shared_receive_relayable.rb:
```diff
@@ -0,0 +1,94 @@
+                                              .and_return(remote_user.encryption_key)
+
+    Workers::ReceiveEncryptedSalmon.new.perform(user.id, generate_xml(entity, remote_user, user))
+    expect(klass.exists?(guid: entity.guid)).to be(false)
```
`be_true` and `be_false` should be things.

cmrd-senya replied:
When I try this instead I get [the following](https://gist.github.com/cmrd-senya/1c1b1d880682ae471b88).

jhass replied:
Meh, nvm then I guess.

Commit:
bump diaspora_federation

cmrd-senya commented:
I realized that some of situations tested here are artificial. In real life, it's not possible, or just doesn't make sense, that sender makes right signature for Salmon and then fails to make a signature for the payload using the same key. Thus, I'll rewrite tests in consideration that we only test situations where the both signatures (Salmon and in-payload, like in relayables) are correct or both signatures are wrong.

Flaburgan commented:
Also consider the situation of a malicious user and pod owner ;)diaspora/diaspora a écritI realized that some of situations tested here are artificial. In real life, it's not possible, or just doesn't make sense, that sender makes right signature for Salmon and then fails to make a signature for the payload using the same key. Thus, I'll rewrite tests in consideration that we only test situations where the both signatures (Salmon and in-payload, like in relayables) are correct or both signatures are wrong.

—Reply to this email directly or view it on GitHub.


  
  


Envoyé depuis Firefox OS

cmrd-senya commented:
Sure, it is considered, it is the only reason why we need signatures at
all :)

ATM we have a situation where entity in payload, for example comment
gets signed by the key of a user, and then packed in Salmon, and Salmon
gets signed with the same key again. There isn't much sense to check
cases, when one signature is ok and the other is broken (moreover, we
probably just don't need these duplicating signatures at all, see #6566).

The fun is that the only case when a signature in the entity is in fact
useful
([this](https://github.com/diaspora/diaspora/pull/6539/files#diff-a1ee075c06bc50b99a8bc09ea96aae48R61)
is the one where we don't check it now :)

I guess that there maybe was some idea behind it, like they wanted, for
example, to make pods exchange messages in P2P style, like pod A posted
a status, pod B fetched the status and resent it to pod C, which resent
it to pod D. That would explain why they duplicated signatures in the
payloads but not only in signed envelopes, and that would explain why
they wanted to keep signatures in the database. But we don't do that, so
most of this signatures code is redundant for us.

Commit:
Allow Retraction to deal with Relayables

Commit:
Fix a security issue that author_signature is not checked on the to-downstream
receive of a federated relayable entity, allowing to forge relayables if you are
an owner of the pod where a parent object is stored.

cmrd-senya commented:
Please, review.

cmrd-senya commented:
diaspora_federation 0.0.9 is released, I guess this one is therefore ready.

jhass commented:
Merged as 4e41b8dc86dceaf43c154e718969e87de3c1a0f6 08b910bd8851839745cdf41fb6d78c9fd8e3baef 922d26f9768b206127b13642f476ca959e9e2dc9 f0fc62e94d27d27e9fc6bc48455b415d555f5922

Thank you!

cmrd-senya commented:
Wow, cool!
